pkg/trace/trace.go

```diff
@@ -0,0 +1,168 @@
+	mutex         *sync.Mutex   // serialize access to the trace directory
+	dir           string        // where stack traces are dumped
+	crashFileGlob string        // pattern to create or match files
+	removeAfter   time.Duration // how long to keep files
+	keepMaxFiles  int           // delete oldest files if there are more than this
+}
+
+var keeper = &traceKeeper{
+	mutex:         &sync.Mutex{},
+	dir:           os.TempDir(),
+	crashFileGlob: "crowdsec-crash.*.txt",
+	removeAfter:   30 * 24 * time.Hour,
+	keepMaxFiles:  100,
+}
+
+func (tk *traceKeeper) list() ([]string, error) {
+	files, err := filepath.Glob(filepath.Join(tk.dir, tk.crashFileGlob))
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+func (tk *traceKeeper) purge() {
+	// Run in a mutex in case of concurrent, recoverable panics from apiserver
+	tk.mutex.Lock()
+	defer tk.mutex.Unlock()
+
+	files, err := tk.list()
+	if err != nil {
+		log.Errorf("error listing crash files for cleanup: %s", err)
+		return
+	}
+
+	// Sort files by modification time, newest first
+	sort.Slice(files, func(i, j int) bool {
+		fi, err := os.Stat(files[i])
+		if err != nil {
+			return true
+		}
+
+		fj, err := os.Stat(files[j])
+		if err != nil {
+			return false
+		}
+
+		return fi.ModTime().After(fj.ModTime())
+	})
+
+	now := time.Now()
+	kept := 0
+
+	for _, file := range files {
+		if kept >= tk.keepMaxFiles {
+			log.Infof("Removing excess trace file: %s", file)
+
+			if err := os.Remove(file); err != nil {
+				log.Errorf("error removing old crash file %s: %s", file, err)
+			}
+
+			continue
+		}
+
+		fi, err := os.Stat(file)
+		if err != nil {
+			log.Errorf("error stating file %s: %s", file, err)
+			continue
+		}
+
+		if now.Sub(fi.ModTime()) > tk.removeAfter {
+			log.Infof("Removing excess trace file: %s", file)
+
+			if err := os.Remove(file); err != nil {
+				log.Errorf("error removing old crash file %s: %s", file, err)
+			}
+
+			continue
+		}
+
+		kept++
+	}
+}
+
+func (tk *traceKeeper) writeStackTrace(iErr any) (string, error) {
+	tk.purge()
+
+	tmpfile, err := os.CreateTemp(tk.dir, tk.crashFileGlob)
+	if err != nil {
+		return "", err
+	}
+	defer tmpfile.Close()
+
+	if _, err := fmt.Fprintf(tmpfile, "error: %+v\n", iErr); err != nil {
+		return "", err
+	}
+
+	if _, err := tmpfile.WriteString(version.FullString()); err != nil {
+		return "", err
+	}
+
+	if _, err := tmpfile.Write(debug.Stack()); err != nil {
+		return "", err
+	}
+
+	return tmpfile.Name(), nil
+}
+
+func (tk *traceKeeper) catchPanic(component string) {
+	r := recover()
+	if r == nil {
+		return
+	}
+
+	log.Errorf("crowdsec - goroutine %s crashed: %s", component, r)
+	log.Error("please report this error to https://github.com/crowdsecurity/crowdsec/issues")
+
+	filename, err := tk.writeStackTrace(r)
+	if err != nil {
+		log.Errorf("unable to write stacktrace: %s", err)
+	}
+
+	log.Errorf("stacktrace/report is written to %s: please join it to your issue", filename)
+	log.Fatal("crowdsec stopped")
```

Comment:
Did we use to fatal at the end of a trace?

Reply:
yes if it's not in apiserver. other goroutines panics are not recoverable and we don't have a return path